src/text_metadata.rs

```diff
@@ -0,0 +1,546 @@
+//!
+//!  let mut writer = encoder.write_header().unwrap();
+//!
+//!  let data = [255, 0, 0, 255, 0, 0, 0, 255]; // An array containing a RGBA sequence. First pixel is red and second pixel is black.
+//!  writer.write_image_data(&data).unwrap(); // Save
+//!
+//!  // We can add a tEXt/zTXt/iTXt at any point before the encoder is dropped from scope. These chunks will be at the end of the png file.
+//!  let tail_ztxt_chunk = ZTXtChunk::new("Comment", "A zTXt chunk after the image data.");
+//!  writer.write_text_chunk(&tail_ztxt_chunk).unwrap();
+//!
+//!  // The fields of the text chunk are public, so they can be mutated before being written to the file.
+//!  let mut tail_itxt_chunk = ITXtChunk::new("Author", "सायंतन खान");
+//!  tail_itxt_chunk.compressed = true;
+//!  tail_itxt_chunk.language_tag = "hi".to_string();
+//!  tail_itxt_chunk.translated_keyword = "लेखक".to_string();
+//!  writer.write_text_chunk(&tail_itxt_chunk).unwrap();
+//!  ```
+
+#![warn(missing_docs)]
+
+use crate::{chunk, encoder, DecodingError, EncodingError};
+use deflate::write::ZlibEncoder;
+use deflate::Compression;
+use encoding::all::{ASCII, ISO_8859_1};
+use encoding::{DecoderTrap, EncoderTrap, Encoding};
+use miniz_oxide::inflate::decompress_to_vec_zlib;
+use std::io::Write;
+
+/// Text encoding errors that is wrapped by the standard EncodingError type
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum TextEncodingError {
+    /// Unrepresentable characters in string
+    Unrepresentable,
+    /// Keyword longer than 79 bytes or empty
+    InvalidKeywordSize,
+    /// Error encountered while compressing text
+    CompressionError,
+}
+
+/// Text decoding error that is wrapped by the standard DecodingError type
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum TextDecodingError {
+    /// Unrepresentable characters in string
+    Unrepresentable,
+    /// Keyword longer than 79 bytes or empty
+    InvalidKeywordSize,
+    /// Missing null separator
+    MissingNullSeparator,
+    /// Compressed text cannot be uncompressed
+    InflationError,
+    /// Using an unspecified value for the compression method
+    InvalidCompressionMethod,
+    /// Using a byte that is not 0 or 255 as compression flag in iTXt chunk
+    InvalidCompressionFlag,
+    /// Missing the compression flag
+    MissingCompressionFlag,
+}
+
+/// A generalized text chunk trait
+pub trait EncodableTextChunk {
+    /// Encode text chunk as Vec<u8> to a `Write`
+    fn encode<W: Write>(&self, w: &mut W) -> Result<(), EncodingError>;
+}
+
+/// Struct representing a tEXt chunk
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TEXtChunk {
+    /// Keyword field of the tEXt chunk. Needs to be between 1-79 bytes when encoded as Latin-1.
+    pub keyword: String,
+    /// Text field of tEXt chunk. Can be at most 2GB.
+    pub text: String,
+}
+
+impl TEXtChunk {
+    /// Constructs a new TEXtChunk.
+    /// Not sure whether it should take &str or String.
+    pub fn new(keyword: &str, text: &str) -> Self {
```

Comment:
Idiomatically, it would be better for this function to take `String`s. That avoids an extra copy if the caller doesn't need the string afterwards